contracts/templates/TemplateLauncher.sol

```diff
@@ -2,77 +2,122 @@
 pragma solidity >=0.6.8;
 import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
-import "../interfaces/ITemplate.sol";
-import "../interfaces/IMesaFactory.sol";
-import "../utils/cloneFactory.sol";
+import "../shared/interfaces/ITemplate.sol";
+import "../shared/interfaces/IMesaFactory.sol";
+import "../shared/utils/cloneFactory.sol";
 
 contract TemplateLauncher is CloneFactory {
     using SafeERC20 for IERC20;
 
-    struct Template {
-        bool exists;
-        uint64 templateId;
-        uint128 index;
-        bool verified;
-    }
-
-    uint256 public templateId;
-    mapping(uint256 => address) private templates;
-    mapping(address => uint256) private templateToId;
-    mapping(address => Template) public templateInfo;
-
-    event TemplateLaunched(address indexed sale, uint256 templateId);
+    event TemplateLaunched(
+        address indexed newTemplate,
+        uint256 templateId,
+        address templateDeployer,
+        string metaData
+    );
     event TemplateAdded(address indexed template, uint256 templateId);
     event TemplateRemoved(address indexed template, uint256 templateId);
     event TemplateVerified(address indexed template, uint256 templateId);
-    event UpdatedTemplateRestriction(bool restrictedTemplates);
+    event TemplateRestrictionUpdated(bool restrictedTemplates);
+    event TemplateMetaDataUpdated(
+        address _launchedTemplate,
+        string _newMetaData
+    );
+
+    mapping(uint256 => address) private template;
+    mapping(address => uint256) private templateToId;
+    mapping(address => bool) public templateVerified;
+
+    struct TemplateData {
+        address deployer;
+        string metaData;
+    }
+
+    mapping(address => TemplateData) public launchedTemplate;
 
+    uint256 templateCounter;
     address public factory;
-    bool public restrictedTemplates = true;
+    bool public restrictedTemplates;
```

Comment:
Let's name this `allowPublicTemplates` or something in that line, so it's a bool flag.